audio/out/ao_sndio.c

```diff
@@ -0,0 +1,349 @@
+    const struct af_to_par *ap;
+    int i;
+
+    p->hdl = sio_open(SIO_DEVANY, SIO_PLAY, 0);
+    if (p->hdl == NULL) {
+        MP_ERR(ao, "can't open sndio %s\n", SIO_DEVANY);
+        goto error;
+    }
+
+    ao->format = af_fmt_from_planar(ao->format);
+
+    sio_initpar(&p->par);
+    for (i = 0, ap = af_to_par;; i++, ap++) {
+        if (i == sizeof(af_to_par) / sizeof(struct af_to_par)) {
+            MP_VERBOSE(ao, "unsupported format\n");
+            p->par.bits = 16;
+            p->par.sig = 1;
+            p->par.le = SIO_LE_NATIVE;
+            break;
+        }
+        if (ap->format == ao->format) {
+            p->par.bits = ap->bits;
+            p->par.sig = ap->sig;
+            if (ap->bits > 8)
+                p->par.le = SIO_LE_NATIVE;
+            if (ap->bits != SIO_BPS(ap->bits))
+                p->par.bps = ap->bits / 8;
+            break;
+        }
+    }
+    p->par.rate = ao->samplerate;
+
+    struct mp_chmap_sel sel = {0};
+    for (int n = 0; n < MP_NUM_CHANNELS+1; n++)
+        mp_chmap_sel_add_map(&sel, &sndio_layouts[n]);
+
+    if (!ao_chmap_sel_adjust(ao, &sel, &ao->channels))
+        goto error;
+
+    p->par.pchan = ao->channels.num;
+    p->par.appbufsz = p->par.rate * 250 / 1000;    /* 250ms buffer */
+    p->par.round = p->par.rate * 10 / 1000;    /*  10ms block size */
+    if (!sio_setpar(p->hdl, &p->par)) {
+        MP_ERR(ao, "couldn't set params\n");
+        goto error;
+    }
+    if (!sio_getpar(p->hdl, &p->par)) {
+        MP_ERR(ao, "couldn't get params\n");
+        goto error;
+    }
+    if (p->par.bps > 1 && p->par.le != SIO_LE_NATIVE) {
+        MP_ERR(ao, "swapped endian output not supported\n");
+        goto error;
+    }
+    if (p->par.bits == 8 && p->par.bps == 1 && !p->par.sig) {
+        ao->format = AF_FORMAT_U8;
+    } else if (p->par.bits == 16 && p->par.bps == 2 && p->par.sig) {
+        ao->format = AF_FORMAT_S16;
+    } else if ((p->par.bits == 32 || p->par.msb) && p->par.bps == 4 && p->par.sig) {
+        ao->format = AF_FORMAT_S32;
+    } else {
+        MP_ERR(ao, "couldn't set format\n");
+        goto error;
+    }
+
+    p->havevol = sio_onvol(p->hdl, volcb, p);
+    sio_onmove(p->hdl, movecb, p);
+
+    p->pfd = calloc (sio_nfds(p->hdl), sizeof (struct pollfd));
+    if (!p->pfd)
+        goto error;
+
+    ao->device_buffer = p->par.bufsz;
+
+    p->buffer = NULL;
+    p->playing = false;
+
+    return 0;
+
+error:
+    if (p->hdl)
+      sio_close(p->hdl);
+
+    return -1;
+}
+
+/*
+ * close device
+ */
+static void uninit(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    if (p->hdl)
+        sio_close(p->hdl);
+
+    free(p->pfd);
+}
+
+/*
+ * stop playing and empty buffers (for seeking/pause)
+ */
+static void reset(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    if (p->playing) {
+        MP_WARN(ao, "Blocking until remaining audio is played... (sndio design bug).\n");
+
+        p->playing = false;
+
+        if (!sio_stop(p->hdl))
```

Comment:
```sio_stop()``` make sense if you have unplayed data.

Reply:
so only if `delay > 0` ?

Reply:
On ```reset()``` buffers must be cleaned, but if they already empty no need to do something.
But this will require more changes to code, to keep that ```sio_start()``` done before ```start()```.
I do this only to minimize ```sio_stop()``` calls, to reduce glitches.

Reply:
> But this will require more changes to code, to keep that sio_start() done before start().
I do this only to minimize sio_stop() calls, to reduce glitches.

From what I understand, `sio_start()` can be called repeatedly without issue (if you run with `SNDIO_DEBUG=2` you can see it detects that it's already started); `sio_stop()` could be another matter. I don't know of any glitches with `sio_stop()`, but it'd be ok to remove it when unnecessary, since its function here is only to block while buffers are drained.